### DISCUSSION TITLE
Advice wanted on Core versus ORM for dynamic schema/table discovery in PostgreSQL

### DISCUSSION THREAD

**Discussion by chrispy-snps** [sqlalchemy/sqlalchemy#11878] [Usage Questions] [answered]
I'm new to databases, SQL, new to PostgreSQL, and new to SQLAlchemy. Let me start off by saying that as a newcomer, SQLAlchemy is indeed indistinguishable from magic!

I'm building a system that works as follows:

* Documents of multiple types are stored in a single database.
* Documents are organized by type into Postgres schemas.
* Each schema can contain one or more tables of documents.
* Each table can have a "context", which is a particular set of fields and values describing the conditions that the documents are valid in.
* Document requests are similarly made with a "context", and all documents from all tables with compatible contexts must be returned.

I have a proof-of-concept working via bare-metal SQL with `psycopg` - schema discovery, table discovery, context compatibility checks, and SQL query construction. The functionality is exactly what we need. But the query construction code is ugly and hard to follow (manipulation of SQL string fragments).

The code would be much cleaner by using SQLAlchemy with its method-chaining magic! But I'm not sure whether to implement this with Core or ORM.

I tried ORM first, but things got complicated because I need to dynamically discover schemas and tables, then dynamically define ORM classes with the correct schema/table/column definitions, then retrieve documents across all the tables in a single multi-schema query (class factories and all that fun stuff). And I think I need to mix Core into my code for the discovery stuff, which adds to the complexity.

Should I push through the complexity and do this with ORM? Or is Core a better fit for dynamic table/schema/query situations where more fine-grained control is needed over schema/table discovery and query construction?

**Comment by CaselIT** [comment, marked as answer]
Hi,

I don't think there is a black or white reply here. A lot will be subjective.

Sqlalchemy can automatically load schemas into tables or orm classes, so that would possibly automate some of the table or class creation.
- Reflect tables: https://docs.sqlalchemy.org/en/20/core/metadata.html#sqlalchemy.schema.MetaData.reflect
- Reflect orm classes https://docs.sqlalchemy.org/en/20/orm/extensions/automap.html#module-sqlalchemy.ext.automap

Assuming the work for your use case, using orm or core becomes a matter of taste mainly / preference.

**Reply by chrispy-snps** [reply]
@CaselIT - thanks Federico! An answer of "it depends" is still an answer, and that is exactly what I was looking for.

I read through the links you provided. In particular, I was not aware of the ORM automapping, so that was quite interesting!

To start, I decided to go with a core-based approach to help me understand SQLAlchemy and database fundamentals. After that, I can evaluate whether a move to ORM will improve the code readability and maintenance for my team.
